modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb

```diff
@@ -61,7 +61,11 @@ def legacy_classes_format(response)
 
   def get_classes(environment)
     future = async_get_classes(environment)
-    future.value!(MAX_CLIENT_TIMEOUT)
+    cache_used = @m.synchronize { !!@etag_cache[environment] } #etags are only available when classes cache is enabled
+    client_timeout = cache_used ? DEFAULT_CLIENT_TIMEOUT : 2 * DEFAULT_CLIENT_TIMEOUT
```

Comment:
Do you know if there are there any reason(s) for keeping the class cache disabled on puppet server? If none, then I'd keep the timeout at a value which keeps foreman somewhat responsive (300 seconds will be pretty unusable, I think) and perhaps expand the warning to include a suggestion to enable the class cache...